### PR TITLE
Implement ConcatFromSequence

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: Implement ConcatFromSequence

**Motivation and Context**
Make ORT opset-11 compliant

### CODE REVIEW

File: onnxruntime/test/providers/cpu/sequence/concat_from_sequence_op_test.cc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  SeqTensors<float> input;
+  input.AddTensor({1, 2}, {0.0f, 1.0f});
+  input.AddTensor({1, 2}, {2.0f, 3.0f});
+  test.AddSeqInput("S", input);
+  test.AddOutput<float>("I", {1, 2, 2}, {0.0f, 2.0f, 1.0f, 3.0f});
+  test.Run();
+}
+
+// TODO:
+// 1. tests for scalar stacking
```
Comment (hariharans29) [inline onnxruntime/test/providers/cpu/sequence/concat_from_sequence_op_test.cc]: Need to add some more tests

File: onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  // number of input tensors in the Sequence to concatenate
+  int input_count = static_cast<int>(X->tensors.size());
+
+  // Hold pointers to the input tensors to be used in the PrepareForCompute() step
+  const auto& input_tensors = X->tensors;
+
+  std::vector<const Tensor*> input_tensor_pointers;
+  input_tensor_pointers.reserve(input_count);
+  for (int i = 0; i < input_count; ++i) {
+    input_tensor_pointers.push_back(std::move(&input_tensors[i]));
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc]: Any reason you need std::move here?

File: onnxruntime/core/providers/cpu/tensor/concat.cc [hunk 3]
```diff
@@ -20,112 +21,146 @@ ONNX_CPU_OPERATOR_KERNEL(
...
+static Status PrepareForCompute(OpKernelContext* ctx, const std::vector<const Tensor*>& input_tensors,
+                                int64_t axis, bool is_stack, Prepare& p) {
+  int input_count = static_cast<int>(input_tensors.size());
+  // Must have atleast one input to concat
   ORT_RETURN_IF_NOT(input_count >= 1, "Must have 1 or more inputs");
-  const Tensor* tensor_pointer = ctx->Input<Tensor>(0);
-  if (tensor_pointer == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
+
+  const Tensor* tensor_pointer = input_tensors[0];
+  if (tensor_pointer == nullptr)
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/concat.cc]: This is not possible since the TensorSeq data structure contains a vector<Tensor> and as long as this vector has a size of non-zero, this ptr can't be NULL. I suggest we do the cardinality check in Compute() using input_tensors.empty()

File: onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  const auto* X = ctx->Input<TensorSeq>(0);
+  ORT_ENFORCE(X != nullptr, "Got nullptr for sequence input.");
+
+  // number of input tensors in the Sequence to concatenate
+  int input_count = static_cast<int>(X->tensors.size());
+
+  // Hold pointers to the input tensors to be used in the PrepareForCompute() step
+  const auto& input_tensors = X->tensors;
+
+  std::vector<const Tensor*> input_tensor_pointers;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc]: Why do we need this separate vector when we already have a vector of Tensor in the TensorSeq struct?

File: onnxruntime/core/providers/cpu/tensor/concat.cc [hunk 5]
```diff
@@ -20,112 +21,146 @@ ONNX_CPU_OPERATOR_KERNEL(
...
-    num_elements *= inputs_0_dims[dimension_index];
+  std::vector<int64_t> output_dims = inputs_0_dims;
+  // 'Concat' mode
+  if (!is_stack) {
+    // While concating, the rank of the output is the same as the input rank(s)
+
+    // Calculate the size of the concatenated axis
+    size_t concat_axis_size = 0;
+    for (int64_t index = 0; index < input_count; index++) {
+      concat_axis_size += input_tensors[index]->Shape()[int(p.axis)];
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/concat.cc]: C++ style cast. static_cast<int>(p.axis)

Comment (hariharans29) [reply onnxruntime/core/providers/cpu/tensor/concat.cc]: Fixed. Thanks.

File: onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc]: I refactored the method in the base class so that it is re-usable for both `Concat` and `ConcatFromSequence` and hence I settled on a middle-ground base class method interface. In `Concat`, we need to use OpKernelContext to parse the inputs and in `ConcatFromSequence`, we get a vector of tensors, and so I thought we could pass in pointers to the input Tensors for both ops. This pointer accumulation should be cheap. Please let me know if you have any concerns.

File: onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc]: Not needed. Removed.

File: onnxruntime/core/providers/cpu/tensor/concat.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/tensor/concat.cc]: Sure, we could remove this check (but many ops have very similar checks and I think this got carried over from the previous implementation)

File: onnxruntime/test/providers/cpu/sequence/concat_from_sequence_op_test.cc (hunk 1 above)
Comment (pranavsharma) [reply onnxruntime/test/providers/cpu/sequence/concat_from_sequence_op_test.cc]: Do you plan to add the tests as part of this PR?

File: onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc (hunk 4 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/sequence/concat_from_sequence.cc]: Alright, let's go with this. We can improve later if desired.

File: onnxruntime/test/providers/cpu/sequence/concat_from_sequence_op_test.cc (hunk 1 above)
Comment (hariharans29) [reply onnxruntime/test/providers/cpu/sequence/concat_from_sequence_op_test.cc]: Oh sorry. I should have added a clarifying statement before resolving.

I did add tests for scalar stacking, stacking with empty inputs, and concatenating a sequence of tensors. Since the base class is being re-used across both Concat and ConcatFromSequence, there is good enough coverage for concatenating.

File: onnxruntime/core/providers/cpu/tensor/concat.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/tensor/concat.cc]: Synced offline - we will do this in a consolidated manner for all ops later.
